[2.0.x] Changes to STM32F1 HAL

victorpv opened the pull request:
Some to correct missing files from previous PR to completely avoid
HardwareTimer Class (reduce overhead). Some changes to formatting.
Changes to DMA ADC to work correctly.
Change to F1 sanity check.

victorpv committed:
Changes to F1 HAL

Some to correct missing files from previous PR to completely avoid
HardwareTimer Class (reduce overhead). Some changes to formatting.
Changes to DMA ADC to work correctly.
Change to F1 sanity check.

thinkyhead committed:
Update HAL_Stm32f1.cpp

thinkyhead commented on Marlin/src/HAL/HAL_STM32F1/HAL_spi_Stm32f1.cpp:
```diff
@@ -79,8 +79,8 @@ void spiBegin() {
   #endif
-  SET_OUTPUT(SS_PIN);
   WRITE(SS_PIN, HIGH);
+  SET_OUTPUT(SS_PIN);
```
Why write to the pin before it's sure to be an output?

AnHardt replied:
I don't know how this is working on a F1. 
On a 8bit AVR, with a externally pulled up pin what was an input, this sequence would mean.
Set the internal pull up additionally - what makes no difference.
Then set the pin to be an output - what drives it high immediately.

The other sequence can have a short moment where the output is low.

AnHardt replied:
If there is no external pull up. Switching on the internal pull up will drive the pin high with low power, Switching it to an output then activates full power.

thinkyhead replied:
Interesting. That's the F1 behavior? And that what is wanted? Is that better for the attached peripheral?

victorpv replied:
In an F1 the ports are input floating by default. If the mode is changed to putput, they will be out push pull, default low.
With the previous sequence, as @AnHardt pointed out there could be a blip of the port being pulled low.
By applying it in this sequence, when the port is set to output it will correctly go straigh from input floating to output high level.
To clarify, the attached peripheral is the SDCard, and this part of the code is called to set up the SPI pins. We don't want CS pulled low at this point. In theory, all the other SPI pins should be in low state and not doing anything, but still the CS pin should not go down unintentionally which is what the previous sequence coud cause.

thinkyhead committed:
Update HAL_timers_Stm32f1.cpp

thinkyhead committed:
Update watchdog_Stm32f1.h

thinkyhead commented on Marlin/src/HAL/HAL_STM32F1/watchdog_Stm32f1.h:
```diff
@@ -39,6 +45,11 @@ void watchdog_init();
-inline void watchdog_reset() { iwdg_feed(); }
+inline void watchdog_reset() {
+  #if PIN_EXISTS(LED)
+    TOGGLE(LED_PIN);  // heart beat indicator
```
I'm not sure if we want this always. Depends on whether users want control over their `LED_PIN` for other uses.

victorpv replied:
Is there a configuration option somewhere to decide what to do with the LED? I did not see any, and copied that bit from the LPC HAL, so the LED would provide a heart bit.

thinkyhead replied:
> Is there a configuration option somewhere to decide what to do with the LED?

No, there is not a config option for it. I'm sure users will ask for one if they need it someday.

> copied that bit from the LPC HAL, so the LED would provide a heart bit.

That's fine, if we're setting a precedent.

thinkyhead committed:
Update watchdog_Stm32f1.h

thinkyhead committed:
Update HAL_timers_Stm32f1.cpp

xC0000005 commented:
Hey, @victorpv, is there a reason the temp timer interval is set to 100? That's causing temp instability in my local build. If I revert that line only (setting it back to 1000), my PID tuning is stable.

victorpv commented:
@xC0000005 I don't remember for sure, but I think I changed it based on what other HALs were doing, with the temp timer having a lower frequency.
I don't think there was any reason other than that, if I get time I'll test it too, although not sure when I will have time. Feel free to submit a PR to change it.
I think I also changed the ADC sampling rate because we were capturing samples way faster than calling the temp timer. Check that to make sure with the new timer frequency you are not reading the same value from the buffer during multiple ISR calls. Likely not, I don't think the ADC capture rate was ever close to mS.
